### PR TITLE
ChibiOS: allow GPIO alt config

### PR DESCRIPTION
**Pull request by IamPete1**
This work is supported by CubePilot.

This allows changing pins to GPIO using `BRD_ALT_CONFIG`. This should work for all the existing types, I have also added support for changing UART RTS and CTS pins to other things. I have not added the ability to change other things to RTS and CTS. 

This also hides `BRD_ALT_CONFIG` on builds where it will not do anything. 

I have implemented this for CubeOrange telem1 RTS CTS. Tested with both relay and button work with `BRD_ALT_CONFIG` = 1 and don't with `BRD_ALT_CONFIG` = 0. 

I have not tested that the RTS CTS pins still work with `BRD_ALT_CONFIG` = 0, any tips on the best way to do what would be appreciated.

### CODE REVIEW

File: libraries/AP_HAL_ChibiOS/UARTDriver.cpp [hunk 1]
```diff
@@ -1605,13 +1605,17 @@ bool UARTDriver::set_options(uint16_t options)
 
 #ifdef HAL_PIN_ALT_CONFIG
     /*
-      allow for RX and TX pins to be remapped via BRD_ALT_CONFIG
+      allow for RX, X, RTS and CTS pins to be remapped via BRD_ALT_CONFIG
```
Comment (IamPete1) [inline libraries/AP_HAL_ChibiOS/UARTDriver.cpp:1608]:
```suggestion
      allow for RX, TX, RTS and CTS pins to be remapped via BRD_ALT_CONFIG
```

**Comment by tridge** [comment]
needs checking with configure_all.py to check that this only impacts boards that where it adds a GPIO alt config

**Comment by IamPete1** [comment]
Tested with: `./Tools/scripts/configure_all.py`, `git add build -f` then reverting the python and CubeHwdef change and  `./Tools/scripts/configure_all.py` again. 

The only diff is the formatting change of a couple of the `PAL_LINE` prints, I went from `PAL_LINE(GPIO%s, %2uU)` to `PAL_LINE(GPIO%s,%uU)` so all the `PAL_LINE` prints are the same and you can easily search for all the `PAL_LINE` uses for a particular pin. 

I think the only other thing is to decide if we should have some convention for ALT GPIO pin numbers, this adds them as 4 and 5 because there the next free on Cube Orange.

**Review by tridge — APPROVED** [review]

**Comment by kniuk** [comment]
How can I check if this feature is available for Matek F765WING ?

**Comment by kniuk** [comment]
@IamPete1 please advice

**Comment by IamPete1** [comment]
@kniuk Yeah, this is only implemented on CubeOrange ATM, but it should just need adding to the hwdef.dat for new boards now it is supported. This it the change for CubeOrange, https://github.com/ArduPilot/ardupilot/pull/18753/commits/387f1b79df868f6104269126bc8e2c696e65ba55 you would need something similar for other boards.

**Comment by kniuk** [comment]
@IamPete1 great! That is what I was thinking of. So I can basically do something like this for MATEKF765WING:
```
# UART7 (telem1)
PE7 UART7_RX UART7
PE8 UART7_TX UART7
PE10 UART7_CTS UART7
PE9  UART7_RTS UART7

+# telem1 RTS and CTS as GPIO in alt config 2 and 3 
+PE10 EXTERN_GPIO1 OUTPUT GPIO(1) ALT(2)
+PE9  EXTERN_GPIO2 OUTPUT GPIO(2) ALT(2)
+PE10 EXTERN_GPIO1 OUTPUT GPIO(1) ALT(3)
+PE9  EXTERN_GPIO2 OUTPUT GPIO(2) ALT(3)
```
I introduced two additional alternative configs, because someone might want both GPIOs and FPORT (ALT_CONFIG=3)
```
# USART6 (RC input), SERIAL7
PC7 TIM3_CH2 TIM3 RCININT FLOAT LOW
PC6 USART6_TX USART6 NODMA

# as an alternative config setup the RX6 pin as an uart. This allows
# for bi-directional UART based receiver protocols such as FPort
# without any extra hardware
PC7 USART6_RX USART6 NODMA ALT(1)
+PC7 USART6_RX USART6 NODMA ALT(3)
```
How are the GPIOs numbered? According to https://ardupilot.org/plane/docs/common-gpios.html they should start from number 50. With above code  would I be able to use RELAY_PIN = 1 to drive PE10 (uart7_cts) pin ? Or  should I check somewhere else for conflicts?

**Comment by IamPete1** [comment]
@kniuk That looks OK, the pwm pins will already be mapped starting at 50, for the Orange I started at 1 it already has GPIO 1 to 3, so the new ones ended up at 4 and 5. Looks like there are no others for MATEKF765WING so I think 1 and 2 makes sense. 

Probably worth checking the GPIO mapping in the generated `hwdef.h` I did not test having the same pins in multiple alt configs.

**Comment by kniuk** [comment]
@IamPete1 so I changed hwdef.dat of MatekF765-Wing:
```
# UART7 (telem1)
PE7 UART7_RX UART7
PE8 UART7_TX UART7
PE10 UART7_CTS UART7
PE9  UART7_RTS UART7

+ # telem1 RTS and CTS as GPIO in alt config 2 and 3 
+ PE10 EXTERN_GPIO1 OUTPUT GPIO(1) ALT(2) #Cts7 pin as GPIO 1 (set RELAY_PINx = 1 to use it)
+ PE9  EXTERN_GPIO2 OUTPUT GPIO(2) ALT(2) #Rts7 pin as GPIO 2 (set RELAY_PINx = 2 to use it)

```
Then I compiled and uploaded firmware to my MatekF765-Wing (my first time :)
I have set:
```
BOARD_ALT_CONFIG = 2
RC6_OPTION = 36 (RELAY4 on/off)  - I have 3 first relays already setup 
RC10_OPTION = 66 (RELAY5 on/off)
RELAY_PIN4 = 1
RELAY_PIN5 = 2
```
I hooked up two LEDs to Rts7 and Cts7 pins on my board. It works no problem. However...
When I tried to use UART7 (SERIAL1_PROTOCOL=2 --MavLink2) I observed some interesting stuff.
As long as both pins are assigned to RELAY_PINx and RELAY_PINy (like above) the communication works no problem. When I don't assign GPIO1 and GPIO2 to any RELAY_PIN (set to -1) MissionPlanner does not connect to my board (via MAVLink WiFi Bridge Version: 1.2.2 on ESP8266 in my case). I fiddled with RELAY_PIN values several times and I am 100% sure about it.

P.S.
I failed to do this (discussed above):
```
PE10 EXTERN_GPIO1 OUTPUT GPIO(1) ALT(2)
PE10 EXTERN_GPIO1 OUTPUT GPIO(1) ALT(3)
```
I got this:
```
Writing hwdef setup in /home/kniuk/ardupilot/build/MatekF765-Wing/hwdef.h
MCU Flags: cortex-m7 ['-mcpu=cortex-m7', '-mfpu=fpv5-d16', '-mfloat-abi=hard']
Error: Duplicate GPIO value 1
Failed to process hwdef.dat ret=1
```

**Pull request merged**